Microservice samples should call `CustomerService.getCustomerInfo` in a global transaction (#57)
scalar-labs/scalardb-samples: spring-data-microservice-transaction-sample/README.md

![ERD](images/ERD.png)

### Service endpoints

The endpoints defined in the services are as follows:

- Customer Service
  - `getCustomerInfo`
  - `getCustomerInfoForTwoPhaseCommit`
  - `payment`
  - `prepare`
  - `validate`
  - `commit`
  - `rollback`
  - `repayment`

- Order Service
  - `placeOrder`
  - `getOrder`
  - `getOrders`

> ### What you can do in this sample application

Author: This section is copied from `microservice-transaction-sample/README.md` to make them consistent.

The sample application supports the following types of transactions:

- Get customer information through the `getCustomerInfo` endpoint of the Customer Service.
- Place an order by using a line of credit through the `placeOrder` endpoint of the Order Service and the `payment`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
  - Checks if the cost of the order is below the customer's credit limit.
  - If the check passes, records the order history and updates the amount the customer has spent.

> - Get order information by order ID through the `getOrder` of the Order Service and `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
> - Get order information by customer ID through the `getOrders` of the Order Service and `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.

Reviewer: ```suggestion
- Get order information by order ID through the `getOrder` endpoint of the Order Service and the `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
- Get order information by customer ID through the `getOrders` endpoint of the Order Service and the `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
```

  - `placeOrder`
  - `getOrder`
  - `getOrders`

### What you can do in this sample application

The sample application supports the following types of transactions:

- Get customer information through the `getCustomerInfo` endpoint of the Customer Service.
- Place an order by using a line of credit through the `placeOrder` endpoint of the Order Service and the `payment`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
  - Checks if the cost of the order is below the customer's credit limit.
  - If the check passes, records the order history and updates the amount the customer has spent.
- Get order information by order ID through the `getOrder` endpoint of the Order Service and the `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
- Get order information by customer ID through the `getOrders` endpoint of the Order Service and the `getCustomerInfoForTwoPhaseCommit`, `prepare`, `validate`, `commit`, and `rollback` endpoints of the Customer Service.
- Make a payment through the `repayment` endpoint of the Customer Service.
  - Reduces the amount the customer has spent.

> Merging the `getCustomerInfoForTwoPhaseCommit` endpoint into the `getCustomerInfo` endpoint may be a good way to reduce the number of endpoints in the Customer Service. We intentionally separated the endpoints to keep them simple as a sample application.

Reviewer: I'd like to put this part into a note section and avoid using the first-person narrative (`we`). How about we change this as follows?:

```suggestion
{% capture notice--info %}
**Note**

Merging the `getCustomerInfoForTwoPhaseCommit` endpoint into the `getCustomerInfo` endpoint may be a good way to reduce the number of endpoints in the Customer Service. The endpoints are intentionally separated for the sake of simplifying them in the sample application.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>
```

Author: @josh-wong Thanks! I'll merge your suggestions. BTW, we are discussing whether to merge `getCustomerInfoForTwoPhaseCommit` into `getCustomerInfo` and this note might be removed later 😅 

> avoid using the first-person narrative (we)

Just out of curiosity, I sometimes see `you` to represent the readers of the document. `you` is okay while `we` should be avoided?

Reviewer: > BTW, we are discussing whether to merge getCustomerInfoForTwoPhaseCommit into getCustomerInfo and this note might be removed later 😅

Sorry, I should have upscrolled to understand that part in more detail😅

> Just out of curiosity, I sometimes see `you` to represent the readers of the document. `you` is okay while `we` should be avoided?

Correct. In documentation, using `you` is preferred when referring to the reader. Using `we` is a little indirect since readers might be confused on if `we` means `Scalar` or if `we` is an indirect way of saying `you`. In this case, we can use passive voice to avoid an awkward sentence construction, like `Scalar intentionally separated ...`.

Author: > Sorry, I should have upscrolled to understand that part in more detail😅

Oh, you don't need to say sorry. Your suggestion apparently improves the current PR and I really appreciate it.

> Using we is a little indirect since readers might be confused on if we means Scalar or if we is an indirect way of saying you

It really makes sense... Thanks for the explanation!

$ docker-compose down
```

## Reference - How the microservice transaction is achieved

> The transactions for placing an order, getting an order and getting orders achieve the microservice transaction, so this section focuses on how the transactions that span the Customer Service and the Order Service are implemented by taking placing an order as an example.

Reviewer: > [!NOTE]
> 
> The following suggestion is the same as the one I mentioned in [microservice-transaction-sample.md](#).

I'd like to clarify this sentence as follows since readers might think `getting an order and getting orders` is a typo. Also, I split this sentence into two since it's become a bit long and includes more details.

```suggestion
The transactions for placing an order, getting a single order, and getting the history of orders achieve the microservice transaction. This section focuses on how the transactions that span the Customer Service and the Order Service are implemented by placing an order as an example.
```

The following sequence diagram shows the transaction for placing an order:

![Sequence Diagram](images/sequence_diagram.png)

### 1. Transaction with a two-phase commit interface is started

When a client sends a request to place an order to the Order Service, `OrderService.placeOrder()` is called, and the microservice transaction starts.

At first, the Order Service starts a transaction with a two-phase commit interface with `ScalarDbTwoPcRepository.executeTwoPcTransaction()` as follows. For reference, see [`OrderService.java`](order-service/src/main/java/sample/order/OrderService.java).

```java
// Start a two-phase commit interface transaction
TwoPcResult<String> result = orderRepository.executeTwoPcTransaction(txId -> {
  ...
}, ...);
```

> The following `CRUD operations are executed`, `Transaction is committed by using the two-phase commit protocol` and `Error handling` are automatically performed by the API.

Reviewer: ```suggestion
The actions in [CRUD operations are executed](#2-crud-operations-are-executed), [Transaction is committed by using the two-phase commit protocol](#3-transaction-is-committed-by-using-the-two-phase-commit-protocol), and [Error handling](#error-handling) are automatically performed by the API.
```

  throw new ScalarDbNonTransientException(message);
}

// Reduce `credit_total` for the customer
customerRepository.update(customer.withCreditTotal(updatedCreditTotal));
```

### 3. Transaction is committed by using the two-phase commit protocol

After the Order Service receives the update that the payment succeeded, the Order Service tries to commit the transaction.

> `ScalarDbTwoPcRepository.executeTwoPcTransaction()` API, called on the Order Service, automatically performs preparations, validations and commits of both the local Order Service and the remote Customer Service. These steps are executed sequentially after the above CRUD operations successfully finish. The implementations to invoke `prepare`, `validate` and `commit` gRPC endpoints of the Customer Service need to be passed as parameters to the API. For reference, see [`OrderService.java`](order-service/src/main/java/sample/order/OrderService.java).

Reviewer: ```suggestion
The `ScalarDbTwoPcRepository.executeTwoPcTransaction()` API, which called on the Order Service, automatically performs preparations, validations, and commits of both the local Order Service and the remote Customer Service. These steps are executed sequentially after the above CRUD operations successfully finish. The implementations to invoke `prepare`, `validate`, and `commit` gRPC endpoints of the Customer Service need to be passed as parameters to the API. For reference, see [`OrderService.java`](order-service/src/main/java/sample/order/OrderService.java).
```

```java
TwoPcResult<PlaceOrderResponse> result = orderRepository.executeTwoPcTransaction(txId ->{
    ...
  },

  Collections.singletonList(
    RemotePrepareCommitPhaseOperations.createSerializable(
      this::callPrepareEndpoint,
      this::callValidateEndpoint,
      this::callCommitEndpoint,
      this::callRollbackEndpoint
    )
  )
);
```

![Sequence Diagram of High Level 2PC API](images/seq-diagram-high-level-2pc-api.png)

> In the `prepare` endpoint of the Customer Service, it resumes and prepares the transaction with `ScalarDbTwoPcRepository.prepareTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).

Reviewer: Does `it` mean ``the `prepare` endpoint``? If so, how about we change this sentence as follows?:

```suggestion
In the `prepare` endpoint of the Customer Service, the endpoint resumes and prepares the transaction by using `ScalarDbTwoPcRepository.prepareTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).
```

```java
customerRepository.prepareTransactionOnParticipant(request.getTransactionId());
```

> In the `validate` endpoint of the Customer Service, it resumes and validates the transaction with `ScalarDbTwoPcRepository.validateTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).

Reviewer: Does `it` mean ``the `validate` endpoint``? If so, how about we change this sentence as follows?:

```suggestion
In the `validate` endpoint of the Customer Service, the endpoint resumes and validates the transaction by using `ScalarDbTwoPcRepository.validateTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).
```

```java
customerRepository.validateTransactionOnParticipant(request.getTransactionId());
```

> In the `commit` endpoint of the Customer Service, it resumes and commits the transaction with `ScalarDbTwoPcRepository.commitTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).

Reviewer: Does `it` mean ``the `commit` endpoint``? If so, how about we change this sentence as follows?:

```suggestion
In the `commit` endpoint of the Customer Service, the endpoint resumes and commits the transaction by using `ScalarDbTwoPcRepository.commitTransactionOnParticipant()`. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).
```

```java
customerRepository.commitTransactionOnParticipant(request.getTransactionId());
```

### Error handling

> If an error happens while executing a transaction, the transaction will be automatically rolled back in both the local Order Service and the remote Customer Service by `ScalarDbTwoPcRepository.executeTwoPcTransaction()`. The implementation to invoke the `rollback` gRPC endpoint of the Customer Service also needs to be passed as a parameter to the API with other ones. For reference, see [`OrderService.java`](order-service/src/main/java/sample/order/OrderService.java).

Reviewer: Edited to avoid passive voice.

```suggestion
If an error happens while executing a transaction, `ScalarDbTwoPcRepository.executeTwoPcTransaction()` will automatically roll back the transaction in both the local Order Service and the remote Customer Service. The implementation to invoke the `rollback` gRPC endpoint of the Customer Service also needs to be passed as a parameter to the API with other ones. For reference, see [`OrderService.java`](order-service/src/main/java/sample/order/OrderService.java).
```

```java
TwoPcResult<PlaceOrderResponse> result = orderRepository.executeTwoPcTransaction(txId ->{
    ...
  },

  Collections.singletonList(
    RemotePrepareCommitPhaseOperations.createSerializable(
      this::callPrepareEndpoint,
      this::callValidateEndpoint,
      this::callCommitEndpoint,
      this::callRollbackEndpoint
    )
  )
);
```

> In the `rollback` endpoint of the Customer Service, it resumes and rolls back the transaction. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).

Reviewer: Does `it` mean ``the `rollback` endpoint``? If so, how about we change this sentence as follows?:

```suggestion
In the `rollback` endpoint of the Customer Service, the endpoint resumes and rolls back the transaction. For reference, see [`CustomerService.java`](customer-service/src/main/java/sample/customer/CustomerService.java).
```

Author: @josh-wong Thanks for the suggestions! I'm committing them.
